File: backend/harness/thinking.py
"""Reasoning-tag handling for thinking GGUF models (Qwen/QwQ/DeepSeek-style).

These models emit chain-of-thought inside <think>…</think> before the answer.
Three failure modes leak reasoning into the visible reply:
  • closed block left in place:  "<think>…</think> answer"
  • generation truncated inside the block (no </think> ever arrives)
  • chat template omits the opening tag, so output is "reasoning</think> answer"
"""

from __future__ import annotations

import re

_TAGS = r"(?:think|thinking|thought|reasoning)"
_OPEN = re.compile(rf"<\s*{_TAGS}\s*>", re.I)
_CLOSE = re.compile(rf"</\s*{_TAGS}\s*>", re.I)
_CLOSED_BLOCK = re.compile(rf"<\s*{_TAGS}\s*>(.*?)</\s*{_TAGS}\s*>", re.I | re.S)


def split_think(text: str) -> tuple[str, str]:
    """Split model output into (visible, reasoning). Both may be empty."""
    if not text:
        return "", ""
    reasoning: list[str] = []
    s = text

    def _cut(m: re.Match) -> str:
        reasoning.append(m.group(1).strip())
        return " "

    s = _CLOSED_BLOCK.sub(_cut, s)

    # Unclosed <think>… — generation stopped while still reasoning
    m = _OPEN.search(s)
    if m:
        reasoning.append(s[m.end() :].strip())
        s = s[: m.start()]

    # Stray </think> with no opener — everything before it is reasoning
    m = _CLOSE.search(s)
    if m:
        reasoning.append(s[: m.start()].strip())
        s = s[m.end() :]

    visible = re.sub(r"\n{3,}", "\n\n", s).strip()
    thoughts = "\n\n".join(p for p in reasoning if p)
    return visible, thoughts


def strip_think(text: str) -> str:
    """Visible part only (empty string if the model never left reasoning)."""
    return split_think(text)[0]

File: frontend/src/lib/thinking.ts
/**
 * Client-side mirror of backend/harness/thinking.py — split streamed model
 * output into (visible, reasoning) so raw <think> text never lands in the
 * chat bubble while tokens stream in.
 */

const TAGS = "(?:think|thinking|thought|reasoning)";
const OPEN = new RegExp(`<\\s*${TAGS}\\s*>`, "i");
const CLOSE = new RegExp(`</\\s*${TAGS}\\s*>`, "i");
const CLOSED_BLOCK = new RegExp(
  `<\\s*${TAGS}\\s*>([\\s\\S]*?)</\\s*${TAGS}\\s*>`,
  "gi"
);

export function splitThink(text: string): { visible: string; reasoning: string } {
  if (!text) return { visible: "", reasoning: "" };
  const reasoning: string[] = [];
  let s = text.replace(CLOSED_BLOCK, (_m, inner: string) => {
    reasoning.push(inner.trim());
    return " ";
  });

  const open = s.match(OPEN);
  if (open && open.index !== undefined) {
    reasoning.push(s.slice(open.index + open[0].length).trim());
    s = s.slice(0, open.index);
  }

  const close = s.match(CLOSE);
  if (close && close.index !== undefined) {
    reasoning.push(s.slice(0, close.index).trim());
    s = s.slice(close.index + close[0].length);
  }

  return {
    visible: s.replace(/\n{3,}/g, "\n\n").trim(),
    reasoning: reasoning.filter(Boolean).join("\n\n"),
  };
}
